design/vela-core/environment_template_appgroup.md

```diff
@@ -0,0 +1,132 @@
+
+
+## Proposal
+
+This design proposes to add the following CRDs to satisfy user requirements.
+
+### 1 Environment CRD
+
+Environment CRD defines the shared-base environment to deploy applications to.
+
+```yaml
+kind: Environment
+spec:
+  clusters: # Apply to the following clusters.
+    - prod-cluster
+
+  secrets: # The secrets that will be created/updated.
+    - name: redis
+      data:
+        url: redis-url
+        password: redis-password
+
+  definitions: # The definitions that will be created/updated.
+    - type: Component
+      name: function
+      source:
+        kube: # using kubectl apply 
+          git: git-repo-url
+          path: path/to/definition # path to a file or a directory
+    - type: Trait
+      name: logging
+      source:
+        helm: # using helm install/upgrade
+          git: git-repo-url
+          path: path/to/chart
+          valuePath: value.yaml
+```
+
+We would add a new env-controller to handle this CR. When a new Environment CR is applied, the env-controller will apply the secrets and definitions to the corresponding clusters. Note that the cluster objects are defined in [v1beta1/cluster_types.go](https://github.com/oam-dev/kubevela/blob/675b0e24db0cedde11c9870242763957b0012f99/apis/core.oam.dev/v1beta1/cluster_types.go#L45-L52).
+
+### 2. ApplicationTemplate CRD
+
+The ApplicationTemplate CRD defines the base template shared to further parameterize or patch more config to.
+
+```yaml
+kind: ApplicationTemplate
+spec:
+  # base template written in CUE
+  template: |
+    output: {
+      apiVersion: core.oam.dev/v1alpha2
+      kind: Application
+      metadata:
+        name: context.name
+      spec:
+        components:
+          - name: instance
+            type: webservice
+            properties:
+              image: test-image
+              cmd:
+                - sleep 100
+            traits:
+              - name: ingress
+          - name: database
+            type: parameter.databaseType
+        environment:
+          namespace: parameter.project
+    }
+    parameter: {
+      project: string | *""
+      databaseType: string | *"rds"
+    }
+```
+
+Note that this is a data object to reference to. It wouldn't have any effects if applied by itself.
+
+### 3. ApplicationSetting CRD
+
+With the above ApplicationTemplate, how could users pick a template and define environment-specific parameters? This is what ApplicationSetting CRD defines.
+
+```yaml
+kind: ApplicationSetting
+spec:
+  applications:
+  - targetEnvs: # names of the Environments to deploy to
+      - dev
+    template: my-app-tpl # name of the ApplicationTemplate
+    config:
+      parameters:
+        project: dev
+        databaseType: mysql
+  - targetEnvs:
+      - prod
+    template: my-app-tpl # name of the ApplicationTemplate
+    config:
+      parameters:
+        project: prod
+        databaseType: rds
+      patch: # kustomize-style overlay patch to base template
```

Comment:
we may only implement the ApplicationTemplate first, and leave the patch feature in the future when users really need it.